homeassistant/components/volvo/coordinator.py

```diff
@@ -0,0 +1,216 @@
+from homeassistant.exceptions import ConfigEntryAuthFailed, HomeAssistantError
+from homeassistant.helpers.device_registry import DeviceInfo
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import DATA_BATTERY_CAPACITY, DOMAIN, MANUFACTURER
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@dataclass
+class VolvoData:
+    """Data for Volvo Cars integration."""
+
+    coordinator: VolvoDataCoordinator
+
+
+type VolvoConfigEntry = ConfigEntry[VolvoData]
+type CoordinatorData = dict[str, VolvoCarsApiBaseModel | None]
+
+
+class VolvoDataCoordinator(DataUpdateCoordinator[CoordinatorData]):
+    """Volvo Data Coordinator."""
+
+    config_entry: VolvoConfigEntry
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        entry: VolvoConfigEntry,
+        api: VolvoCarsApi,
+    ) -> None:
+        """Initialize the coordinator."""
+
+        super().__init__(
+            hass,
+            _LOGGER,
+            config_entry=entry,
+            name=DOMAIN,
+            update_interval=timedelta(seconds=135),
+        )
+
+        self.api: VolvoCarsApi = api
+
+        self.vehicle: VolvoCarsVehicle
+        self.device: DeviceInfo
+
+        # The variable is set during _async_setup().
+        self._refresh_conditions: dict[
+            str, tuple[Callable[[], Coroutine[Any, Any, Any]], bool]
+        ] = {}
+
+    async def _async_setup(self) -> None:
+        """Set up the coordinator.
+
+        This method is called automatically during
+        coordinator.async_config_entry_first_refresh.
+        """
+        _LOGGER.debug("%s - Setting up", self.config_entry.entry_id)
+
+        try:
+            vehicle = await self.api.async_get_vehicle_details()
+        except VolvoAuthException as ex:
+            raise ConfigEntryAuthFailed(
+                translation_domain=DOMAIN,
+                translation_key="unauthorized",
+                translation_placeholders={"message": ex.message},
+            ) from ex
+
+        if vehicle is None:
+            raise HomeAssistantError(
+                translation_domain=DOMAIN, translation_key="no_vehicle"
+            )
+
+        self.vehicle = vehicle
+        self.data = {}
+
+        device_name = (
+            f"{MANUFACTURER} {vehicle.description.model} {vehicle.model_year}"
+            if vehicle.fuel_type == "NONE"
+            else f"{MANUFACTURER} {vehicle.description.model} {vehicle.fuel_type} {vehicle.model_year}"
+        )
+
+        self.device = DeviceInfo(
+            identifiers={(DOMAIN, vehicle.vin)},
+            manufacturer=MANUFACTURER,
+            model=f"{vehicle.description.model} ({vehicle.model_year})",
+            name=device_name,
+            serial_number=vehicle.vin,
+        )
+
+        self.hass.config_entries.async_update_entry(
+            self.config_entry,
+            title=f"{MANUFACTURER} {vehicle.description.model} ({vehicle.vin})",
+        )
```

Comment:
Let's set it once during config flow and then the user can change to whatever they want.